### DISCUSSION TITLE
How does the createMiddleware function work with parsedInput?

### DISCUSSION THREAD

**Discussion by palmmymumu** [next-safe-action/next-safe-action#286] [Q&A]
For example, in a case where you need to check if ctx.user.userId has permission to access the data of the payment specified in parsedInput.payId, and **return next({ ctx: { payment } })**. to next action

**Comment by TheEdoRan** [comment]
The `next` function returns an object with information about the action execution. If you `await` it, you'll see that it returns `parsedInput`, you can then access `parsedInput.payId` and then return the result to complete the execution, or stop it there.

You can see an example in the middleware page of the docs, by looking at the [logging](https://next-safe-action.dev/docs/define-actions/middleware#instance-level-middleware) one.

**Reply by palmmymumu** [reply]
I'm creating middleware guard to protect resources from being access/edit/execution, If I await next() for parsedInput will it already been executed? (e.g. in case of editing protected data)

**Reply by TheEdoRan** [reply]
> If I await next() for parsedInput will it already been executed?

Yep, so not suitable. You can instead check the `clientInput` (middleware argument) to process the `payId`. This is unparsed data though, because in next-safe-action input/output validation is performed **after** middleware functions.

**Reply by AndreMaz** [reply]
Hey @TheEdoRan I want to enrich the data by adding `createdById` | `updatedById` | `deletedById` to the payload. 
However, as you've mentioned `clientInput` is not suited for this and the `parsedInput` is not available in the middleware.

So the question to you, what's the most elegant way of achieving this? 

I guess that I could do something like
```ts
export const createAsset = actionClient
  .metadata({
    serviceName: "asset",
    actionName: "createAsset",
    roles: ["tenant-admin"],
    opType: "create",
  })
  .schema(
    z.object({
      data: z.object({ entry: assetCreateSchema }),
      meta: z.object({
        orgId: z.string().uuid(),
      }),
    }),
  )
  .action(async ({ metadata: { createdById }, parsedInput: { data } }) => {
    // modify data to include createdById
    data.entry.createdById = createdById;
    return await insertAssetHandler({ data });
  });
  ```
  
But this would mean that I would need to update all the actions that I have. I would prefer a "centralized" approach where the middleware handles data enrichment.

Do you have any alternative?

**Reply by AndreMaz** [reply]
Hey @TheEdoRan just saw that you're working on v8. Is there any change of adding additional middlware/hook that's executed **after** the current [middleware](https://next-safe-action.dev/docs/define-actions/middleware) but **before** the `action()` handler?

This would allow me "enrich" the data just like mentioned in my original post :point_up_2:
